### DISCUSSION TITLE
Windows Runtime

### DISCUSSION THREAD

**Discussion by fafalone** [twinbasic/twinbasic#1728] [General]
Microsoft seems to be essentially deprecating the regular Windows API VBx/tB programmers have always relied on for most non-trivial programs.

Virtually every new feature is locked behind Windows RT now. I think it's worth exploring a way for tB to access these classes. It's currently *possible* to use *some of* it through extremely unfriendly vtable calling; I've looked into defining the interfaces as traditional COM but some of them it seems it just isn't possible, and all of them are so incredibly obfuscated I couldn't get any practical coverage in a reasonable time. Obviously external DLLs are a solution. C linkage is another potential solution (though that's still difficult, C++ at least would be needed to be practical). But then why not just develop in that other language?

How possible is it to bind with these classes in an automatic way without needing to hand-rewrite tens of thousands of classes that take at least 30-60min each?

**Comment by WaynePhillipsEA** [comment]
For the uninitiated, could you give some examples of the problematic interfaces?

**Reply by fafalone** [reply]
```
                    virtual HRESULT STDMETHODCALLTYPE add_Dismissed(
                        __FITypedEventHandler_2_Windows__CUI__CNotifications__CToastNotification_Windows__CUI__CNotifications__CToastDismissedEventArgs* handler,
                        EventRegistrationToken* token
                        ) = 0;
```

argument relevant defs:

```
typedef ITypedEventHandler<ABI::Windows::UI::Notifications::ToastNotification*, ABI::Windows::UI::Notifications::ToastDismissedEventArgs*> __FITypedEventHandler_2_Windows__CUI__CNotifications__CToastNotification_Windows__CUI__CNotifications__CToastDismissedEventArgs_t;
#define __FITypedEventHandler_2_Windows__CUI__CNotifications__CToastNotification_Windows__CUI__CNotifications__CToastDismissedEventArgs ABI::Windows::Foundation::__FITypedEventHandler_2_Windows__CUI__CNotifications__CToastNotification_Windows__CUI__CNotifications__CToastDismissedEventArgs_t
```

--->

```
namespace ABI { namespace Windows { namespace Foundation {
template <>
struct __declspec(uuid("61c2402f-0ed0-5a18-ab69-59f4aa99a368"))
ITypedEventHandler<ABI::Windows::UI::Notifications::ToastNotification*, ABI::Windows::UI::Notifications::ToastDismissedEventArgs*> : ITypedEventHandler_impl<ABI::Windows::Foundation::Internal::AggregateType<ABI::Windows::UI::Notifications::ToastNotification*, ABI::Windows::UI::Notifications::IToastNotification*>, ABI::Windows::Foundation::Internal::AggregateType<ABI::Windows::UI::Notifications::ToastDismissedEventArgs*, ABI::Windows::UI::Notifications::IToastDismissedEventArgs*>>
{
    static const wchar_t* z_get_rc_name_impl()
    {
        return L"Windows.Foundation.TypedEventHandler`2<Windows.UI.Notifications.ToastNotification, Windows.UI.Notifications.ToastDismissedEventArgs>";
    }
};
```

```
namespace ABI {
    namespace Windows {
        namespace UI {
            namespace Notifications {
                class ToastNotification;
            } /* Notifications */
        } /* UI */
    } /* Windows */
} /* ABI */
```

Now buried under all that there is a normal interface, but it doesn't seem it's asking for it directly and I have no idea how you'd supply whatever it does want.

```
                MIDL_INTERFACE("3f89d935-d9cb-4538-a0f0-ffe7659938f8")
                IToastDismissedEventArgs : public IInspectable
                {
                public:
                    virtual HRESULT STDMETHODCALLTYPE get_Reason(
                        ABI::Windows::UI::Notifications::ToastDismissalReason* value
                        ) = 0;
                };
```

All I wanted to do is show a simple flyout notification, and I've had to run down dozens of slightly less obfuscated things until I got stuck here.

But the bigger problem is even if could tell me how to decipher this, it's just not practical to rewrite it all to be consumed under current abilities. It would remain a novelty even more impractical for real world use than drivers, which, especially if that file not found issue was ever solved for minifilters, is definitely within reach of practical application for real world use. WinRT currently isn't, IMHO.

You can see VBF member -Franky-'s existing WinRT use in VB6 in the VBC_UwP_* downloads on https://www.activevb.de/cgi-bin/upload/catalog.pl

**Reply by fafalone** [reply]
Ok here's a better example, in the OCR classes... results are provided as a collection of IOcrLine interfaces:

```
[propget] HRESULT Lines([out] [retval] Windows.Foundation.Collections.IVectorView<Windows.Media.Ocr.OcrLine*>** value);
```

IVectorView, which has GetAt and GetSize methods to supply collection members, is not a single interface. It's methods are defined only in terms of pure template code. It's seemingly generated on the fly from:

```
namespace ABI { namespace Windows { namespace Foundation { namespace Collections {
template <>
struct __declspec(uuid("60c76eac-8875-5ddb-a19b-65a3936279ea"))
IVectorView<ABI::Windows::Media::Ocr::OcrLine*> : IVectorView_impl<ABI::Windows::Foundation::Internal::AggregateType<ABI::Windows::Media::Ocr::OcrLine*, ABI::Windows::Media::Ocr::IOcrLine*>>
{
    static const wchar_t* z_get_rc_name_impl()
    {
        return L"Windows.Foundation.Collections.IVectorView`1<Windows.Media.Ocr.OcrLine>";
    }
};
// Define a typedef for the parameterized interface specialization's mangled name.
// This allows code which uses the mangled name for the parameterized interface to access the
// correct parameterized interface specialization.
typedef IVectorView<ABI::Windows::Media::Ocr::OcrLine*> __FIVectorView_1_Windows__CMedia__COcr__COcrLine_t;
#define __FIVectorView_1_Windows__CMedia__COcr__COcrLine ABI::Windows::Foundation::Collections::__FIVectorView_1_Windows__CMedia__COcr__COcrLine_t
```

and 

```
    template <typename T> struct abi<winrt::Windows::Foundation::Collections::IVectorView<T>>
    {
        struct __declspec(novtable) type : inspectable_abi
        {
            virtual int32_t __stdcall GetAt(uint32_t, arg_out<T>) noexcept = 0;
            virtual int32_t __stdcall get_Size(uint32_t*) noexcept = 0;
            virtual int32_t __stdcall IndexOf(arg_in<T>, uint32_t*, bool*) noexcept = 0;
            virtual int32_t __stdcall GetMany(uint32_t, uint32_t, arg_out<T>, uint32_t*) noexcept = 0;
        };
    };
```

A "no vtable" dynamic interface.

That GUID is *only* for IOcrLine-providing IVectorView. Other IVectorView uses have different GUIDs.

So we're required to build individual IVectorViews for the many, many places it's used. Assuming the idea is to make it usable via friendly typelibs, as opposed to -Franky-'s code that just gets the pointer and calls it with DispCallFunc, where you don't need to know the GUID.

**Comment by Xumaquer** [comment]
I've managed to use a little of WinRT APIs in VB6 (mostly just a winrt messagebox for testing purposes), but i had to manually dig through lots of IDL files i've found on github
there's also the problem of some COM interfaces used on winRT relying on templates like IVectorView,IIterable,etc.
I've mostly written them using a IDL file so i only had to make some small changes for it to work (as i mostly found their declaration on IDL files that i found on github) 

I think the "legit" way to bind them  would be to make a IDL compiler that compiles IDL files using MIDL 3.0 into a TLB or some similar format (i think the MIDL compiler doesn't compile MIDL 3.0 to a TLB) that TB can read, so that it can generate bindings to those objects from the IDL files , which is actually how microsoft does the bindings to winrt objects for C++,C#,javascript and VB.NET. They are all generated from IDL files using the MIDL compiler

Also the way WinRT objects are created is different, we usually need to use functions like RoActivateInstance or RoGetActivationFactory
instead of CoCreateInstance

**Comment by fafalone** [comment]
It's not a complete enough library to be used for anything, but here's what I was starting with trying to make an app to show those modern flyout notifications ("Toast notifications"). 

[WinRTLib.zip](https://github.com/twinbasic/twinbasic/files/13644519/WinRTLib.zip)

**Reply by Xumaquer** [reply]
That's cool but just a tip, i've saw that windows is keeping the implementation of WinRT related functions like Ro functions and HSTRING related functions in "combase.dll" but apparently that could "change in the future" so they're recommending people to use "API sets" instead

https://learn.microsoft.com/en-us/windows/win32/apiindex/windows-apisets#api-set-contract-names

with Ro functions being on the API set : api-ms-win-core-winrt-l1-1-0
and HSTRING functions being on the API set : api-ms-win-core-winrt-string-l1-1-0
